utils/selection.py

```diff
@@ -0,0 +1,588 @@
+
+    Raises:
+        NoSelectionElements: If no choices provided
+        SelectionCancelled: If user cancels or times out
+    """
+    if len(choices) == 0:
+        raise NoSelectionElements()
+    elif len(choices) == 1 and not force_select:
+        return choices[0]
+
+    # Use specialized monster selection if requested
+    if is_monster and pm:
+        return await select_monster_with_dm_feedback(ctx=ctx, choices=choices, key=key, query=query, timeout=timeout)
+
+    def create_embed(page: int) -> disnake.Embed:
+        """Create selection embed for given page."""
+        pages = paginate(choices, 10)
+        current_choices = pages[page] if page < len(pages) else []
+
+        description = ""
+        if query:
+            description += f"Your input was: `{query}`\n"
+        description += "Which one were you looking for? (Type the number or `c` to cancel)\n"
+        if len(pages) > 1:
+            description += "`n` to go to the next page, or `p` for previous\n"
+        description += "\n"
+
+        for i, choice in enumerate(current_choices):
+            global_index = i + 1 + page * 10
+            description += f"[{global_index}] - {key(choice)}\n"
+
+        description += "\n**Instructions**\n"
+        if not pm:
+            description += "Use buttons below OR Type your choice in this channel."
+        else:
+            description += f"Use buttons below OR Type your choice in {ctx.channel.mention}. This message was PMed to you to hide the monster name."
+
+        if message:
+            description += f"\n\n**Note**\n{message}"
+
+        embed = disnake.Embed(title="Multiple Matches Found", description=description, colour=0x36393F)
+
+        if len(pages) > 1:
+            embed.set_footer(text=f"Page {page + 1}/{len(pages)}")
+
+        return embed
+
+    def create_text_check():
+        """Create message check function for text input."""
+
+        def text_check(msg):
+            if msg.author != ctx.author or msg.channel != ctx.channel:
+                return False
+            content = msg.content.lower().strip()
+            if content in ("c", "n", "p"):
+                return True
+            try:
+                choice_num = int(content)
+                return 1 <= choice_num <= len(choices)
+            except ValueError:
+                return False
+
+        return text_check
+
+    # Send initial message with buttons
+    page = 0
+    pages = paginate(choices, 10)
+    embed = create_embed(page)
+    view = StatelessSelectionView(choices, page, key, query or "")
+
+    if pm:
+        embed.add_field(name="Instructions", value=f"Click buttons or type in {ctx.channel.mention}", inline=False)
+        select_msg = await ctx.author.send(embed=embed, view=view)
+    else:
+        select_msg = await ctx.send(embed=embed, view=view)
+
+    # Dual input handling loop - buttons OR text
+    text_check = create_text_check()
+    current_timeout = timeout
+
+    while True:
```

Comment:
`while True` is an instant-red-flag for me. Some boundary condition would be better. Maybe a maximum amount of times to receive an event or while the interaction is not expired. Open to options here, though.